File: Java/src/_0930_Binary_Subarrays_With_Sum/Solution.java
package _0930_Binary_Subarrays_With_Sum;

import java.util.HashMap;

/**
 * Date: 2021/7/8 9:29
 * Content:
 * 执行用时：17 ms, 在所有 Java 提交中击败了71.62%的用户
 * 内存消耗：43.3 MB, 在所有 Java 提交中击败了9.54%的用户
 */
public class Solution {

    public int numSubarraysWithSum(int[] nums, int goal) {
        int n = nums.length;
        int[] s = new int[n + 1];
        for (int i = 1; i <= n; i++) s[i] = s[i - 1] + nums[i - 1];

        int res = 0;
        HashMap<Integer, Integer> cnt = new HashMap<>();
        for (int i = 0; i <= n; i++) {
            res += cnt.getOrDefault(s[i] - goal, 0);
            cnt.put(s[i], cnt.getOrDefault(s[i], 0) + 1);
        }
        return res;
    }

    public static void main(String[] args) {

        int[] nums = {1, 0, 1, 0, 1};
        System.out.println((new Solution()).numSubarraysWithSum(nums, 2));  // 4
    }
}

File: Python/_0930_Binary_Subarrays_With_Sum/Solution.py
# coding=utf-8
# Date: 2021/7/8 9:31
from typing import List


# 执行用时：284 ms, 在所有 Python3 提交中击败了36.55%的用户
# 内存消耗：18.6 MB, 在所有 Python3 提交中击败了14.22%的用户
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        n = len(nums)
        s = [0 for _ in range(n + 1)]
        for i in range(1, n + 1):
            s[i] = s[i - 1] + nums[i - 1]

        res = 0
        cnt = {}
        for i in range(1, n + 1):
            if s[i - 1] not in cnt:
                cnt[s[i - 1]] = 0
            cnt[s[i - 1]] += 1
            if s[i] - goal in cnt:
                res += cnt[s[i] - goal]
        return res


if __name__ == "__main__":
    print(Solution().numSubarraysWithSum([1, 0, 1, 0, 1], 2))  # 4
